src/SwedbankPay.Episerver.Checkout/SwedbankPayCheckoutGateway.cs

```diff
@@ -42,15 +59,81 @@ public PaymentProcessingResult ProcessPayment(IOrderGroup orderGroup, IPayment p
         {
-            OrderGroup = payment.Parent.Parent;
-            _orderForm = payment.Parent;
-            return ProcessPayment(payment, ref message);
+            return ProcessPayment(payment as IPayment, ref message);
         }
 
         public bool ProcessPayment(IPayment payment, ref string message)
         {
+            if (_orderForm == null)
+            {
+                _orderForm = OrderGroup.GetFirstForm();
+            }
+
+            _shipment = _orderForm.Shipments.FirstOrDefault();
+
+            var result = AsyncHelper.RunSync(() => ProcessPayment(payment, _shipment));
+            message = result.Message;
+
+            return result.Status;
+        }
+
+
+        /// <summary>Process payment associated with shipment.</summary>
+        /// <param name="payment">The payment</param>
+        /// <param name="shipment">The shipment</param>
+        /// <param name="message">The message</param>
+        /// <returns>True if process successful, otherwise false</returns>
+        public bool ProcessPayment(Payment payment, Shipment shipment, ref string message)
+        {
+            var result = AsyncHelper.RunSync(() => ProcessPayment(payment as IPayment, shipment));
+            message = result.Message;
+            return result.Status;
+        }
+
+        /// <summary>Process payment associated with shipment.</summary>
+        /// <param name="orderGroup">The order group.</param>
+        /// <param name="payment">The payment.</param>
+        /// <param name="shipment">The shipment.</param>
+        /// <returns><c>True</c> if process successful, otherwise <c>False</c>.</returns>
+        public PaymentProcessingResult ProcessPayment(IOrderGroup orderGroup, IPayment payment, IShipment shipment)
+        {
+            OrderGroup = orderGroup;
+            _orderForm = orderGroup.GetFirstForm();
+            var result = AsyncHelper.RunSync(() => ProcessPayment(payment, shipment));
+            var message = result.Message;
+            return result.Status
+                ? PaymentProcessingResult.CreateSuccessfulResult(message)
+                : PaymentProcessingResult.CreateUnsuccessfulResult(message);
+        }
+
+        public async Task<PaymentStepResult> ProcessPayment(IPayment payment, IShipment shipment)
+        {
+            var paymentStepResult = new PaymentStepResult();
+            _shipment = shipment;
+
+            if (_orderForm == null)
+            {
+                _orderForm = (payment as Payment)?.Parent ?? OrderGroup?.Forms.FirstOrDefault(form => form.Payments.Contains(payment));
+            }
+            if (OrderGroup == null)
+            {
+                OrderGroup = (_orderForm as OrderForm)?.Parent;
+            }
+
+            if (OrderGroup == null)
+            {
+                paymentStepResult.Message = "OrderGroup is null";
+                throw new Exception(paymentStepResult.Message);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/gh/SwedbankPay/swedbank-pay-episerver-checkout/pullRequest?prid=5328029)